src/Grpc.NetCore.HttpClient/Internal/GrpcCall.cs

```diff
@@ -0,0 +1,433 @@
+        {
+            var message = CreateHttpRequestMessage();
+            SetMessageContent(request, message);
+            StartSend(client, message);
+        }
+
+        public void StartClientStreaming(System.Net.Http.HttpClient client)
+        {
+            var message = CreateHttpRequestMessage();
+            ClientStreamWriter = CreateWriter(message);
+            StartSend(client, message);
+        }
+
+        public void StartServerStreaming(System.Net.Http.HttpClient client, TRequest request)
+        {
+            var message = CreateHttpRequestMessage();
+            SetMessageContent(request, message);
+            StartSend(client, message);
+            ClientStreamReader = new HttpContentClientStreamReader<TRequest, TResponse>(this);
+        }
+
+        public void StartDuplexStreaming(System.Net.Http.HttpClient client)
+        {
+            var message = CreateHttpRequestMessage();
+            ClientStreamWriter = CreateWriter(message);
+            StartSend(client, message);
+            ClientStreamReader = new HttpContentClientStreamReader<TRequest, TResponse>(this);
+        }
+
+        public void Dispose()
+        {
+            if (!Disposed)
+            {
+                Disposed = true;
+
+                _callCts.Cancel();
+                _ctsRegistration?.Dispose();
+                _writerCtsRegistration?.Dispose();
+                _deadlineTimer?.Dispose();
+                HttpResponse?.Dispose();
+                ClientStreamReader?.Dispose();
+                ClientStreamWriter?.Dispose();
+
+                _callCts.Dispose();
+            }
+        }
+
+        public void EnsureNotDisposed()
+        {
+            if (Disposed)
+            {
+                throw new ObjectDisposedException(nameof(GrpcCall<TRequest, TResponse>));
+            }
+        }
+
+        public void EnsureHeadersValid()
+        {
+            if (_headerValidationError != null)
+            {
+                throw new InvalidOperationException(_headerValidationError);
+            }
+        }
+
+        public Exception CreateCanceledStatusException()
+        {
+            if (_headerValidationError != null)
+            {
+                return new InvalidOperationException(_headerValidationError);
+            }
+
+            var statusCode = DeadlineReached ? StatusCode.DeadlineExceeded : StatusCode.Cancelled;
+            return new RpcException(new Status(statusCode, string.Empty));
+        }
+
+        public void FinishResponse()
+        {
+            if (ResponseFinished)
+            {
+                return;
+            }
+
+            ResponseFinished = true;
+
+            // Clean up call resources once this call is finished
+            // Call may not be explicitly disposed when used with unary methods
+            // e.g. var reply = await client.SayHelloAsync(new HelloRequest());
+            Dispose();
+
+            var status = GetStatusCore(HttpResponse);
```

Comment:
Still using a disposed object here, what's wrong with moving this before `Dispose`

Reply:
The only property being used from the HttpResponse is trailers which dispose doesn't effect. However, there is no reason status couldn't be fetched before the dispose and then an error is thrown afterwards.